From acf128fb8870ffede3da5c86260857f592f1f095 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20L=C3=B6ffler?= <st.loeffler@gmail.com>
Date: Sun, 5 Nov 2017 20:53:42 +0100
Subject: [PATCH 1/2] Only check for Type1 fonts in custom directory if path is
 non-NULL

Otherwise, programs using poppler may crash
Proposed upstream at https://bugs.freedesktop.org/show_bug.cgi?id=49037
---
 poppler/GlobalParamsWin.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/poppler/GlobalParamsWin.cc b/poppler/GlobalParamsWin.cc
index d4eb5871..2056df18 100644
--- a/poppler/GlobalParamsWin.cc
+++ b/poppler/GlobalParamsWin.cc
@@ -382,7 +382,7 @@ void GlobalParams::setupBaseFonts(const char *dir)
 
         GooString *fontName = new GooString(displayFontTab[i].name);
 
-        if (dir) {
+        if (dir && displayFontTab[i].t1FileName) {
             GooString *fontPath = appendToPath(new GooString(dir), displayFontTab[i].t1FileName);
             if (FileExists(fontPath->c_str()) || FileExists(replaceSuffix(fontPath, ".pfb", ".pfa")->c_str())) {
                 addFontFile(fontName, fontPath);
-- 
2.31.1


From bd766857f74b5cf133c7b480af481d58029ae57d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20L=C3=B6ffler?= <st.loeffler@gmail.com>
Date: Mon, 6 Nov 2017 06:59:37 +0100
Subject: [PATCH 2/2] Use a custom font directory on Windows

Uses the share/fonts directory alongside the application/poppler library
---
 poppler/GlobalParams.cc    | 32 +++++++++++++++++++++++++++++++-
 poppler/GlobalParamsWin.cc |  2 +-
 2 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/poppler/GlobalParams.cc b/poppler/GlobalParams.cc
index 01975901..d3569abb 100644
--- a/poppler/GlobalParams.cc
+++ b/poppler/GlobalParams.cc
@@ -150,6 +150,36 @@ static const char *get_poppler_datadir(void)
 #    undef POPPLER_DATADIR
 #    define POPPLER_DATADIR get_poppler_datadir()
 
+static const char *
+get_poppler_fontsdir (void)
+{
+  static char retval[MAX_PATH];
+  static int beenhere = 0;
+
+  unsigned char *p;
+
+  if (beenhere)
+    return retval;
+
+  if (!GetModuleFileName (hmodule, (CHAR *) retval, sizeof(retval) - 20))
+    return POPPLER_DATADIR;
+
+  p = _mbsrchr ((unsigned char *) retval, '\\');
+  *p = '\0';
+  p = _mbsrchr ((unsigned char *) retval, '\\');
+  if (p) {
+    if (stricmp ((const char *) (p+1), "bin") == 0)
+      *p = '\0';
+  }
+  strcat (retval, "\\share\\fonts");
+
+  beenhere = 1;
+
+  return retval;
+}
+#undef POPPLER_FONTSDIR
+#define POPPLER_FONTSDIR get_poppler_fontsdir ()
+
 #endif
 
 //------------------------------------------------------------------------
@@ -845,7 +875,7 @@ GooString *GlobalParams::findFontFile(const GooString *fontName)
 {
     GooString *path = nullptr;
 
-    setupBaseFonts(nullptr);
+    setupBaseFonts(POPPLER_FONTSDIR);
     globalParamsLocker();
     const auto fontFile = fontFiles.find(fontName->toStr());
     if (fontFile != fontFiles.end()) {
diff --git a/poppler/GlobalParamsWin.cc b/poppler/GlobalParamsWin.cc
index 2056df18..c0f760e8 100644
--- a/poppler/GlobalParamsWin.cc
+++ b/poppler/GlobalParamsWin.cc
@@ -504,7 +504,7 @@ GooString *GlobalParams::findSystemFontFile(const GfxFont *font, SysFontType *ty
     if (!fontName)
         return nullptr;
     std::unique_lock<std::recursive_mutex> locker(mutex);
-    setupBaseFonts(nullptr);
+    setupBaseFonts(POPPLER_FONTSDIR);
 
     // TODO: base14Name should be changed?
     // In the system using FontConfig, findSystemFontFile() uses
-- 
2.31.1

